### PR TITLE
C# patterns reference: property pattern

### PR DESCRIPTION
**Pull request by pkulikov**
Contributes to #20584

### CODE REVIEW

File: docs/csharp/language-reference/operators/patterns.md [hunk 1]
```diff
@@ -110,6 +112,24 @@ As the preceding example shows, you can repeatedly use the pattern combinators i
 
 :::code language="csharp" source="snippets/patterns/PatternCombinators.cs" id="WithParentheses":::
 
+## Property pattern
+
+Beginning with C# 8.0, you use a *property pattern* to match expression's properties or fields against nested patterns, as the following example shows:
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/patterns.md:117]: nit:

```suggestion
Beginning with C# 8.0, you use a *property pattern* to match an expression's properties or fields against nested patterns, as the following example shows:
```

File: docs/csharp/language-reference/operators/patterns.md [hunk 2]
```diff
@@ -110,6 +112,24 @@ As the preceding example shows, you can repeatedly use the pattern combinators i
...
+
+:::code language="csharp" source="snippets/patterns/PropertyPattern.cs" id="BasicExample":::
+
+A property pattern matches an expression when an expression result is non-null and every nested pattern matches the corresponding property or field of the expression result.
+
+You can also add a runtime type check and a variable declaration to a property pattern, as the following example shows:
+
+:::code language="csharp" source="snippets/patterns/PropertyPattern.cs" id="WithTypeCheck":::
+
+As you can use any pattern as a nested pattern in a property pattern, you can perform *recursive* pattern matching. That is, you can match parts of data against nested patterns, as the following example shows:
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/patterns.md:127]: ping @MadsTorgersen 

As we keep working on this, I want to consistent in our use of "nested" and "recursive" patterns. I don't think we are now, and I'm not confident in the distinction.

**Review by BillWagner — COMMENTED** [review]
This looks great @pkulikov 

I've got a question out for Mads, and one small suggestion, and then it's ready.

Note: We had a technical issue with builds yesterday. It has been resolved, so any updated commit should build correctly.

File: docs/csharp/language-reference/operators/patterns.md (hunk 2 above)
Comment (pkulikov) [reply docs/csharp/language-reference/operators/patterns.md:127]: I understand that in the following way: 1. "recursive" refers to the "pattern matching" combo, not to "pattern". "Recursive pattern" doesn't sound to me yet. 2. Nested patterns can be also called as "subpatterns".

Comment (MadsTorgersen) [reply docs/csharp/language-reference/operators/patterns.md:127]: Here's how I've been thinking of it:
- _Recursive patterns_ are a term for the _kinds_ of patterns that contain other patterns; e.g. positional, property, relational and logical patterns.
- _Nested patterns_ are concrete occurrences in code where a pattern is nested within another pattern.

I.e. "recursive pattern" classifies a language feature, "nested pattern" describes code.

I don't have deeply held convictions here. That's just where my mind goes to with these terms.

File: docs/csharp/language-reference/operators/patterns.md [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
-ms.date: 03/10/2021
+ms.date: 03/16/2021
 helpviewer_keywords: 
   - "pattern matching [C#]"
 ---
 # Patterns (C# reference)
 
 Intro.
 
+With [property](#property-pattern) and positional patterns, you can perform *recursive* pattern matching. That is, to match parts of data against nested patterns, or sub-patterns.
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/patterns.md:12]: Rewording based on Mads' definitions:

```suggestion
[Property](#property-pattern) and positional patterns support *recursive* pattern matching. That is, to match parts of data against nested patterns, or sub-patterns.
```

File: docs/csharp/language-reference/operators/patterns.md (hunk 2 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/patterns.md:127]: This paragraph matches Mads' definitions (or my current understanding). This looks great.

**Review by BillWagner — COMMENTED** [review]
Thanks again @pkulikov 

I made one suggestion based on Mads' comment to consider "recursive patterns" as a language feature, and "nested patterns" as code that uses recursive patterns.

See what you think. When we agree, I'm ready to merge this.

**Comment by pkulikov** [comment]
@BillWagner I've rewritten the paragraphs about recursive pattern matching. Please check the latest commit.

**Review by BillWagner — APPROVED** [review]
I like where this landed @pkulikov 

I'm ready to :shipit: when you are.

**Comment by pkulikov** [comment]
@BillWagner thank you. I'm ready as well. Let's merge it.

**Pull request merged**